### PR TITLE
kernel: spinlock: k_spin_is_locked introduction

### PR DESCRIPTION
**Pull request by amr-sc**
Currently spinlock internals are directly accessed from the tests. This way the test becomes bound to the particular spinlock implementation. To remove this unnecessary dependency the distinct API to check if spinlock is locked is introduced.

k_spin_is_locked should be used for the spinlock testing only, so the scope of this API is intentionally restricted.

### CODE REVIEW

**Review by evgeniy-paltsev — DISMISSED** [review]

**Review by npitre — DISMISSED** [review]

**Comment by amr-sc** [comment]
Hi @evgeniy-paltsev, sorry for the disturbance!
I've made a small clean-up for the comment, so your approval was lost.

**Review by evgeniy-paltsev — DISMISSED** [review]

**Review by nashif — DISMISSED** [review]

**Comment by amr-sc** [comment]
@andyross , @nashif , @npitre , @evgeniy-paltsev , guys, I'm not authorized to merge this PR. Could you please help?

**Review by cfriedt — DISMISSED** [review]

File: include/zephyr/spinlock.h
Comment (kartben) [inline include/zephyr/spinlock.h:271]: This is not properly documented since k_spin_is_locked symbol is not set when building doxygen documentation.  The comment block ends up applying to K_SPINLOCK_BREAK further down this file.

image

Option 1:

```suggestion
#if defined(CONFIG_SMP) && defined(CONFIG_TEST) || defined(__DOXYGEN__)
```

Option 2:

> k_spin_is_locked should be used for the spinlock testing only, so the scope of this API is intentionally restricted.

Should this be internal then?

**Review by kartben — DISMISSED** [review]

File: include/zephyr/spinlock.h
Comment (amr-sc) [reply include/zephyr/spinlock.h:271]: Hi @kartben , nice catch, thanks! I think both your suggestions make sense

**Review by andyross — COMMENTED** [review]
Some nitpicks & notes to consider. 

Not sure if I really see the need here though.  Subsystem tests are totally allowed to whitebox access to their subsystem, it would be extremely hard to write some test cases without doing that.  I mean, if you caught tests/subsys/net/toast_butter_protocol poking at spinlock internals, that would be one thing.  But tests/kernel/spinlock?  Totally fine by my book.

If you really want this to be a test-only hook, it probably shouldn't be named "k_" like the kernel APIs are.  The current scheme is to name such internal-but-cross-subsystem symbols "z_", though it seems like that may be changing back to a bare underscore soon, IIRC.

I'm going to be very mildly passive aggressive and withhold a +1, but I won't stand in the way if you can get someone else on board. :)

File: include/zephyr/spinlock.h
Comment (evgeniy-paltsev) [reply include/zephyr/spinlock.h:275]: Hi @andyross,

> Not sure why TEST is being tested here? If it's an API it should be an API for everyone. Also if you put the #if's inside the function and not outside, you don't need to special-case doc generation (it just turns into an empty inline for everyone else).

We introduce `k_spin_is_locked` only for the testing purposes - so we can check spinlock state in tests without usage of spinlock structure internal fields. This is usefull abstration as in https://github.com/zephyrproject-rtos/zephyr/pull/61541 we adding another spinlock implementation with different internal fields.

However I don't see any valid use-case of this API in real application - there always be a race conditions - even if we try to use it with interrupts disabled (as other core still may lock / unlock this spinlock).

That's why we have `CONFIG_TEST` guard - to avoid using this API in real applications.

**Comment by npitre** [comment]
Suggestion: put the comment inside the #ifdef and remove the `/**` marker 
by using only `/*`. This is for internal testing purposes so this doesn't 
have to be part of the auto-generated documentation.

**Review by npitre — APPROVED** [review]

**Comment by kartben** [comment]
> Suggestion: put the comment inside the #ifdef and remove the `/**` marker by using only `/*`. This is for internal testing purposes so this doesn't have to be part of the auto-generated documentation.

FWIW, Doxygen already ignores anything prefixed with `z_`, but good call nevertheless :)

https://github.com/zephyrproject-rtos/zephyr/blob/1eab46e36dc21e3eed53335f9b1cfda8189fe2c0/doc/zephyr.doxyfile.in#L1015-L1018

File: include/zephyr/spinlock.h
Comment (amr-sc) [reply include/zephyr/spinlock.h:278]: IMHO this is probably an overkill, we shouldn't consider k_spin_is_locked for usage in real applications. This is just a hook for white-box testing that should help to seamlessly switch between the spinlock implementations without a need for test modifications. IMHO, in that sense there is no need to overload its semantics.

**Review by npitre — APPROVED** [review]

**Review by evgeniy-paltsev — APPROVED** [review]

File: include/zephyr/spinlock.h
Comment (cfriedt) [reply include/zephyr/spinlock.h:278]: Technically, you could declare the function only in a Ztest header 🤷‍♂️

Comment (npitre) [reply include/zephyr/spinlock.h:278]: That would be a bad idea. This is preparation for the ticket-based spinlock PR which would provide an alternative implementation for this. Better if implementation detail alternatives remain in the same file.

**Review by npitre — APPROVED** [review]

**Review by peter-mitsis — APPROVED** [review]

**Comment by amr-sc** [comment]
@dcpleung , @galak , @carlescufi , @nashif , @MaureenHelm , guys, could you please take a look at k_spin_is_locked API ? Thanks!

**Comment by carlescufi** [comment]
@amr-sc I was waiting for @andyross' approval to merge this. Will ping him on Discord.

**Comment by andyross** [comment]
>  @amr-sc I was waiting for @andyross' approval to merge this. Will ping him on Discord.

Heh.  My position remains a very mildly passive-aggressive "I don't see the need, but it's harmless".   So no +1, but go right ahead.

**Pull request merged**
